omnixai/explainers/ranking/agnostic/validity.py

```diff
@@ -0,0 +1,215 @@
+#
+# Copyright (c) 2022 salesforce.com, inc.
+# All rights reserved.
+# SPDX-License-Identifier: BSD-3-Clause
+# For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
+#
+"""
+Ranking Explainer for tabular data.
+"""
+import numpy as np
+from typing import Callable, List
+from ...base import ExplainerBase
+from ....data.tabular import Tabular
+import scipy
+import itertools
+import pandas as pd
+
+from ....explanations.ranking.agnostic.validity import ValidExplanation
+
+
+class ValidityRankingExplainer(ExplainerBase):
+    """
+    Ranking Explainer for Tabular Data.
+    """
+
+    explanation_type = "local"
+    alias = ["rank"]
+
+    def __init__(
+        self,
+        training_data: Tabular,
+        features: List,
+        predict_function: Callable,
+        preprocessing_function: Callable = None,
+    ):
+        """
+        :param training_data: The data used to initialize a Ranking explainer. ``training_data``
+            can be the training dataset for training the machine learning model.
+        :param features: The list of features to be explained by the valid per-query algorithm
+        :param predict_function: The prediction function corresponding to the model to explain.
+            the outputs of the ``predict_function`` are the document scores.
+        :param preprocessing_function: The preprocessing function to process/transform the feature attributes
+        before prediction. The output of the preprocessing function is the input to the ranking model.
+        """
+        super().__init__()
+        self.training_data = training_data.to_pd()
+        self.features = features
+        self.mean_features = self._compute_stats(
+            self.training_data, self.features, "mean"
+        )
+        self.median_features = self._compute_stats(
+            self.training_data, self.features, "median"
+        )
+
+        if preprocessing_function:
+            self.preprocessing_fn = preprocessing_function
+        else:
+            self.preprocessing_fn = lambda x: x.to_pd()
+        self.predict_fn = predict_function
+
+    @staticmethod
+    def _compute_stats(data: pd.DataFrame, features: List, mask_type):
+        if mask_type == "mean":
+            stats = data.mean()
+        if mask_type == "median":
+            stats = data.median()
+
+        for f in features:
+            if f not in stats:
+                most_common_value = data[f].value_counts().idxmax()
+                stats[f] = most_common_value
+
+        return stats
+
+    def _compute_mask(self, mask, idx):
+        if mask == "mean":
+            return self.mean_features[self.features[idx]]
+        elif mask == "median":
+            return self.median_features[self.features[idx]]
+        else:
+            return 0.0
+
+    @staticmethod
+    def _create_copy(sample):
```

Comment:
A simpler way to do this is using "copy" package

Reply:
Done